src/libopensc/card-openpgp.c

```diff
@@ -2283,46 +2371,123 @@ pgp_calculate_and_store_fingerprint(sc_card_t *card, time_t ctime,
 
 	LOG_FUNC_CALLED(card->ctx);
 
-	/* temporary workaround: protect v3 cards against non-RSA */
-	if (key_info->algorithm != SC_OPENPGP_KEYALGO_RSA)
-		LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+	/* constructing public-key packet length */
+	/* RSA */
+	if (key_info->algorithm == SC_OPENPGP_KEYALGO_RSA) {
 
-	if (modulus == NULL || exponent == NULL || mlen == 0 || elen == 0) {
-		sc_log(card->ctx, "Null data (modulus or exponent)");
-		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
-	}
+		if (key_info->u.rsa.modulus == NULL 
+		   || key_info->u.rsa.exponent == NULL
+		   || (key_info->u.rsa.modulus_len >> 3) == 0
+		   || (key_info->u.rsa.exponent_len >> 3) == 0) {
+			sc_log(card->ctx, "Null data (modulus or exponent)");
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+		}
 
-	/* http://tools.ietf.org/html/rfc4880  page 41, 72 */
-	pk_packet_len =   1   /* version number */
-	                + 4   /* creation time */
-	                + 1   /* algorithm */
-	                + 2   /* algorithm-specific fields: RSA modulus+exponent */
-	                + mlen
-	                + 2
-	                + elen;
+		/* https://tools.ietf.org/html/rfc4880  page 41, 72 */
+		pk_packet_len =   1   /* version number */
+				+ 4   /* creation time */
+				+ 1   /* algorithm */
+				+ 2   /* algorithm-specific fields: RSA modulus+exponent */
+				+ (key_info->u.rsa.modulus_len >> 3)
+				+ 2
+				+ (key_info->u.rsa.exponent_len >> 3);
+
+	}
+	/* ECC */
+	else if (key_info->algorithm == SC_OPENPGP_KEYALGO_ECDH
+		|| key_info->algorithm == SC_OPENPGP_KEYALGO_ECDSA) {
+		if (key_info->u.ec.ecpoint == NULL || (key_info->u.ec.ecpoint_len) == 0) {
+			sc_log(card->ctx, "Null data (modulus or exponent)");
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+		}
+
+		/* https://tools.ietf.org/html/rfc4880  page 41, 72 
+		 * and https://tools.ietf.org/html/rfc6637 section 9 (page 8 and 9) */
+		pk_packet_len =   1   /* version number */
+				+ 4   /* creation time */
+				+ 1   /* algorithm */
+				+ 1   /* oid len */
+				+ (key_info->u.ec.oid_len) /* oid */
+				+ (key_info->u.ec.ecpoint_len); /* ecpoint */
+
+		/* KDF parameters for ECDH */
+		if (key_info->algorithm == SC_OPENPGP_KEYALGO_ECDH) {
+			/* https://tools.ietf.org/html/rfc6637#section-8 */
+			pk_packet_len +=   1	/* number of bytes */
+					 + 1	/* version number */
+					 + 1	/* KDF algo */
+					 + 1;	/* KEK algo */
+		}
+	}
+	else
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+	sc_log(card->ctx, "pk_packet_len is %lu", pk_packet_len);
 
 	fp_buffer_len = 3 + pk_packet_len;
 	p = fp_buffer = calloc(fp_buffer_len, 1);
 	if (!p) {
 		LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_ENOUGH_MEMORY);
 	}
 
+	/* constructing public-key packet */
```

Comment:
Using `sc_asn1_encode` would avoid manual encoding...

Reply:
I try to refactor the code these days. But it will probably take some time for me.

Reply:
I am sorry, but I really don't know how to use `sc_asn1_encode` and don't feel comfortable with it. You may can help me (e.g. give some advice or hint how it was done similarly elsewhere) or you are fine with keeping it the way it was the last years.

Reply:
nevermind, most of it isn't ASN.1